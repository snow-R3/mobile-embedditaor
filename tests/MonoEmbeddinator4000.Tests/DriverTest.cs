using System.CodeDom.Compiler;
using System.IO;
using System.Linq;
using System.Text;
using CppSharp.Generators;
using NUnit.Framework;

namespace MonoEmbeddinator4000.Tests
{
    /// <summary>
    /// A set of integration tests / approval tests verifying generated java code
    /// </summary>
    [TestFixture]
    public class DriverTest
    {
        string temp = Path.Combine(Path.GetTempPath(), "hello.dll");
        Project project;
        Options options;
        Driver driver;

        const string HelloWorldSource = @"
namespace Example { 
    public class Hello { 
        public string World() { 
            return ""Hello, World!""; 
        } 
    } 
}";

        [TearDown]
        public void TearDown()
        {
            if (File.Exists(temp))
                File.Delete(temp);
        }

        void RunFromSource(string sourceCode)
        {
            var parameters = new CompilerParameters
            {
                OutputAssembly = temp,
            };
            AssemblyGenerator.CreateFromSource(sourceCode, parameters);

            project = new Project();
            options = new Options
            {
                GeneratorKind = GeneratorKind.Java,
            };

            options.OutputDir =
                project.OutputPath = Path.GetDirectoryName(GetType().Assembly.Location);
            project.Assemblies.Add(temp);

            driver = new Driver(project, options);
            Assert.IsTrue(driver.Run(), "Call to Driver.Run() failed!");
        }

        [Test]
        public void Empty()
        {
            driver = new Driver(new Project(), new Options
            {
                GeneratorKind = GeneratorKind.Java,
            });
            Assert.IsTrue(driver.Run()); //This runs but doesn't throw
        }

        [Test]
        public void HelloFiles()
        {
            RunFromSource(HelloWorldSource);

            var builder = new StringBuilder();
            foreach (var file in driver.Output.Files.Keys)
            {
                builder.AppendLine(file);
            }
            Approvals.Verify(builder.ToString());
        }

        [Test]
        public void Hello_java()
        {
            RunFromSource(HelloWorldSource);

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            string java = File.ReadAllText(path);
            Approvals.Verify(java);
        }

        [Test]
        public void Native_hello_dll_java()
        {
            RunFromSource(HelloWorldSource);

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.Last());
            string java = File.ReadAllText(path);
            Approvals.Verify(java);
        }

        [Test]
        public void UpperCase_java()
        {
            RunFromSource(@"
namespace Example { 
    public class HELLO { 
        public string WORLD() { 
            return ""Hello, World!""; 
        } 
    } 
}");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            string java = File.ReadAllText(path);
            Approvals.Verify(java); //TODO: I don't know if "String wORLD()" is what we want
        }
    }
}
