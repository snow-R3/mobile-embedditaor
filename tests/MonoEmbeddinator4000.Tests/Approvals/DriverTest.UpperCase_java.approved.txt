/**
 * ----------------------------------------------------------------------------
 * <auto-generated>
 * This is autogenerated code by Embeddinator-4000.
 * Do not edit this file or all your changes will be lost after re-generation.
 * </auto-generated>
 * ----------------------------------------------------------------------------
 */
package hello_dll.example;

import mono.embeddinator.*;
import com.sun.jna.*;

public class HELLO {
    public com.sun.jna.Pointer __object;

    public HELLO(com.sun.jna.Pointer object) { this.__object = object; }

    public HELLO() {
        __object = hello_dll.Native_hello_dll.INSTANCE.Example_HELLO_new();
        mono.embeddinator.Runtime.checkExceptions();
    }

    public String wORLD() {
        String __ret = hello_dll.Native_hello_dll.INSTANCE.Example_HELLO_WORLD(__object);
        mono.embeddinator.Runtime.checkExceptions();
        return __ret;
    }
}
